pkg/hostpath/controllerserver.go

```diff
@@ -139,3 +162,227 @@ func (cs *controllerServer) ValidateVolumeCapabilities(ctx context.Context, req
+	// requested sourceVolumeId and sourceVolumeId of snapshot that has been created.
+	if exSnap, err := getSnapshotByName(req.GetName()); err == nil {
+		// Since err is nil, it means the snapshot with the same name already exists need
+		// to check if the sourceVolumeId of existing snapshot is the same as in new request.
+		if exSnap.VolID == req.GetSourceVolumeId() {
+			// same snapshot has been created.
+			return &csi.CreateSnapshotResponse{
+				Snapshot: &csi.Snapshot{
+					Id:             exSnap.Id,
+					SourceVolumeId: exSnap.VolID,
+					CreatedAt:      exSnap.CreateAt,
+					SizeBytes:      exSnap.SizeBytes,
+					Status:         exSnap.Status,
+				},
+			}, nil
+		}
+		return nil, status.Error(codes.AlreadyExists, fmt.Sprintf("snapshot with the same name: %s but with different SourceVolumeId already exist", req.GetName()))
+	}
+
+	volumeID := req.GetSourceVolumeId()
+	hostPathVolume, ok := hostPathVolumes[volumeID]
+	if !ok {
+		return nil, status.Error(codes.Internal, "volumeID is not exist")
+	}
+
+	snapshotID := uuid.NewUUID().String()
+	createAt := time.Now().UnixNano()
+	volPath := hostPathVolume.VolPath
+	file := snapshotRoot + snapshotID + ".tgz"
+	args := []string{"czf", file, "-C", volPath, "."}
+	executor := utilexec.New()
+	out, err := executor.Command("tar", args...).CombinedOutput()
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("failed create snapshot: %v: %s", err, out))
+	}
+
+	fileInfo, err := os.Stat(file)
+	if err != nil {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("failed create snapshot: %v", err))
+	}
+
+	glog.V(4).Infof("create volume snapshot %s", file)
+	snapshot := hostPathSnapshot{}
+	snapshot.Name = req.GetName()
+	snapshot.Id = snapshotID
+	snapshot.VolID = volumeID
+	snapshot.Path = file
+	snapshot.CreateAt = createAt
+	snapshot.SizeBytes = fileInfo.Size()
+	snapshot.Status = &csi.SnapshotStatus{
+		Type:    csi.SnapshotStatus_READY,
+		Details: fmt.Sprint("Successfully create snapshot"),
+	}
+
+	hostPathVolumeSnapshots[snapshotID] = snapshot
+
+	return &csi.CreateSnapshotResponse{
+		Snapshot: &csi.Snapshot{
+			Id:             snapshot.Id,
+			SourceVolumeId: snapshot.VolID,
+			CreatedAt:      snapshot.CreateAt,
+			SizeBytes:      snapshot.SizeBytes,
+			Status:         snapshot.Status,
+		},
+	}, nil
+}
+
+func (cs *controllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteSnapshotRequest) (*csi.DeleteSnapshotResponse, error) {
+	// Check arguments
+	if len(req.GetSnapshotId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Snapshot ID missing in request")
+	}
+
+	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_CREATE_DELETE_SNAPSHOT); err != nil {
+		glog.V(3).Infof("invalid delete snapshot req: %v", req)
+		return nil, err
+	}
+	snapshotID := req.GetSnapshotId()
+	glog.V(4).Infof("deleting volume %s", snapshotID)
+	path := snapshotRoot + snapshotID + ".tgz"
+	os.RemoveAll(path)
```

Comment:
Could log if the file not file for some reason. May be a good idea to return error if not found, which would be good for e2e tests.